core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java

```diff
@@ -20,9 +20,19 @@
 package org.elasticsearch.common.inject;
 
 /**
+ * This interface can be added to a Module to spawn sub modules. DO NOT USE.
  *
+ * This is fundamentally broken.
+ * <ul>
+ * <li>If you have a plugin with multiple modules, return all the modules at once.</li>
+ * <li>If you are trying to make the implementation of a module "pluggable", don't do it.
+ * This is not extendable because custom implementations (using onModule) cannot be
+ * registered before spawnModules() is called.</li>
+ * </ul>
+ * @deprecated Create modules directly
  */
-public interface SpawnModules {
+@Deprecated
```

Comment:
I'll be very happy if we find ourselves in a place where we simply don't need to replace a module. I didn't check the full code base to validate that indeed that works. I just know that I had to extend a module once for testability (i.e. exposing the extensibility for tests only, without exposing the extensibility for plugins). 

In any case, that doesn't change the reasoning behind properly modularizing the code base such that, 1) each module is the only entity responsible for its sub modules, 2) modules can be extensible at any level